sa/db-next/boulder_sa/20230919000000_RevokedCertificates.sql

```diff
@@ -0,0 +1,21 @@
+-- +migrate Up
+-- SQL in section 'Up' is executed when this migration is applied
+
+CREATE TABLE `revokedCertificates` (
+  `id` bigint(20) NOT NULL AUTO_INCREMENT,
+  `issuerID` bigint(20) NOT NULL,
+  `serial` varchar(255) NOT NULL,
+  `notAfter` datetime NOT NULL,
+  `shardIdx` bigint(20) NOT NULL,
+  `revokedDate` datetime NOT NULL,
+  `revokedReason` int(11) NOT NULL,
+  PRIMARY KEY (`id`),
+  KEY `notAfter_shardIdx_idx` (`notAfter`, `shardIdx`)
```

Comment:
After some conversation with @jcjones I learned that even if the `DATETIME` comes second, it can still cause the index to be too big, which could cause the query planner to go to disk instead of using the index. For our purposes, truncating this value to an hour would be sufficient since we really only care about when we get to stop including a row in CRLs (or serving 404 for OCSP).

And yeah putting the `issuerID` in there is a good idea IMO.

Reply:
The cardinality of the index is highly important, but I don't know for certain which ordering here is best, and I suspect the only way to find out will be to run synthetic benchmarks with a few million rows in a MariaDB.

That said, indexes on full-precision `datetime` types are problematical. We do it everywhere, and everywhere it hurts performance. We need to start doing better, and for a fresh table with need for such an index, we should draw a line.

Reply:
The latest patchset replaces the `notAfter` column with `notAfterHour`. It rounds the notAfter _up_ at insert time to ensure we don't think certs expire before they actually do, and conversely rounds the notAfter _down_ at select time, to ensure we're getting more certs than we strictly need to.

I've also added the issuerID to the index, so the index now covers all three columns used in the select's WHERE clause, in order of least-to-most-specific: issuerID cuts the selection space by 1/2, shardIdx cuts the selection space by 1/128, and notAfterHour cuts the selection space by 1/(24*90).